queries/reflect.go

```diff
@@ -441,18 +405,69 @@ func getBoilTag(field reflect.StructField) (name string, recurse bool) {
 	return nameFragment, true
 }
 
-func makeColsKey(typ reflect.Type, cols []string) colBindingKey {
+var (
+	mappingCachesMu sync.Mutex
+	mappingCaches   = make(map[reflect.Type]*mappingCache)
+)
+
+func getMappingCache(typ reflect.Type) *mappingCache {
+	mappingCachesMu.Lock()
+	defer mappingCachesMu.Unlock()
+
+	cache := mappingCaches[typ]
+	if cache != nil {
+		return cache
+	}
+
+	cache = newMappingCache(typ)
+	mappingCaches[typ] = cache
+
+	return cache
+}
+
+type mappingCache struct {
+	typ reflect.Type
+
+	mu          sync.Mutex
+	structMap   map[string]uint64
+	colMappings map[string][]uint64
+}
+
+func newMappingCache(typ reflect.Type) *mappingCache {
+	return &mappingCache{
+		typ:         typ,
+		structMap:   MakeStructMapping(typ),
+		colMappings: make(map[string][]uint64),
+	}
+}
+
+func (b *mappingCache) mapping(cols []string) ([]uint64, error) {
 	buf := strmangle.GetBuffer()
+	defer strmangle.PutBuffer(buf)
+
 	for _, s := range cols {
 		buf.WriteString(s)
+		buf.WriteByte(0)
```

Comment:
I'm concerned about this line. One of the reasons to write the string is to ensure that the order of the columns themselves is correct, not just the number of columns.

`id, username` and `username, id` as columns would now cause a cache key collision and could potentially result in the mapping returned being pointers to the wrong things.

Reply:
I'm confused by this; this code is identical to the original code except that I've added NUL between the columns: https://github.com/volatiletech/sqlboiler/blob/0116414cbb4d642dfef515aeec1e54ebf60317ea/queries/reflect.go#L447

I'm still writing the strings, in the same order.

I can remove the NUL to preserve the original logic, but I added it because one could technically have three columns "foo", "bar", and "foobar", and two queries asking for `["foo", "bar"]` and `["foobar"]` could collide if the column names are coming in unquoted. But that's already true now.

Reply:
I can remove it and look into submitting a PR with it and a regression test, if that's preferred, but I'm definitely still writing the column names in the same order.

Reply:
I'm confused by my own comment. For some reason I imagined a `-` in front of that line. My bad! This is indeed an improvement.

Reply:
Thanks for re-looking!